File: src/java/main/src/main/java/com/tencent/wcdb/orm/Field.java
// Created by qiuwenchen on 2023/3/30.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.wcdb.orm;

import com.tencent.wcdb.winq.Column;
import com.tencent.wcdb.winq.Schema;

import java.io.File;

public class Field<T> extends Column {
    protected TableBinding<T> binding = null;

    public TableBinding<T> getTableBinding() {
        return binding;
    }
    protected int fieldId = 0;

    protected String name;

    public int getFieldId() {
        return fieldId;
    }

    private boolean isAutoIncrement = false;

    public boolean isAutoIncrement() {
        return isAutoIncrement;
    }

    public Field(String name, TableBinding<T> binding, int fieldId, boolean isAutoIncrement) {
        super(name, binding.baseBinding().getBaseBinding());
        this.name = name;
        this.binding = binding;
        this.fieldId = fieldId;
        this.isAutoIncrement = isAutoIncrement;
    }

    protected Field() {
        super();
    }

    protected Field<T> copySelf() {
        Field<T> field = new Field<>();
        field.cppObj = copy(cppObj);
        field.name = name;
        field.fieldId = fieldId;
        field.isAutoIncrement = isAutoIncrement;
        field.binding = binding;
        return field;
    }

    public String getName() {
        return name;
    }

    public Field<T> table(String tableName) {
        Field<T> field = copySelf();
        field.inTable(field.cppObj, tableName);
        return field;
    }

    @Override
    public Field<T> of(String schema) {
        Field<T> field = copySelf();
        Schema newSchema = new Schema(schema);
        field.inSchema(field.cppObj, newSchema.getCppObj());
        return field;
    }

    @Override
    public Field<T> of(Schema schema) {
        Field<T> field = copySelf();
        field.inSchema(field.cppObj, schema.getCppObj());
        return field;
    }

    public static String getTableName(Field field) {
        assert field != null;
        if(field != null) {
            assert field.getTableBinding().bindingTableName() != null;
            return field.getTableBinding().bindingTableName();
        }
        return null;
    }

    public static String getTableName(Field[] fields) {
        assert fields != null && fields.length > 0;
        Field field = fields[0];
        return getTableName(field);
    }

    public static <T> TableBinding<T> getBinding(Field<T> field) {
        assert field != null;
        if(field != null) {
            assert field.getTableBinding() != null;
            return field.getTableBinding();
        }
        return null;
    }

    public static <T> TableBinding<T> getBinding(Field<T>[] fields) {
        assert fields != null && fields.length > 0;
        Field field = fields[0];
        return getBinding(field);
    }
}

File: src/rust/wcdb_core/src/orm/field.rs
use crate::base::cpp_object::CppObjectTrait;
use crate::orm::table_binding::TableBinding;
use crate::winq::column::Column;
use crate::winq::identifier::IdentifierTrait;
use std::ffi::c_void;

pub struct Field<T> {
    column: Column,
    name: String,
    binding: *const dyn TableBinding<T>,
    field_id: usize,
    is_auto_increment: bool,
    is_primary_key: bool,
}

impl<T> CppObjectTrait for Field<T> {
    fn set_cpp_obj(&mut self, _cpp_obj: *mut c_void) {
        self.column.set_cpp_obj(_cpp_obj);
    }

    fn get_cpp_obj(&self) -> *mut c_void {
        self.column.get_cpp_obj()
    }

    fn release_cpp_object(&mut self) {
        self.column.release_cpp_object();
    }
}

impl<T> IdentifierTrait for Field<T> {
    fn get_description(&self) -> String {
        self.column.get_description()
    }
}

impl<T> Field<T> {
    pub fn new(
        name: &str,
        binding: *const dyn TableBinding<T>,
        field_id: usize,
        is_auto_increment: bool,
        is_primary_key: bool,
    ) -> Field<T> {
        let bind = unsafe { &*binding };
        Field {
            column: Column::new_with_binding(name, bind.base_binding().get_base_binding()),
            name: name.to_string(),
            binding,
            field_id,
            is_auto_increment,
            is_primary_key,
        }
    }

    pub fn get_column(&self) -> &Column {
        &self.column
    }

    pub fn get_name(&self) -> &str {
        &self.name
    }

    pub fn get_field_id(&self) -> usize {
        self.field_id
    }

    pub fn get_table_binding(&self) -> &dyn TableBinding<T> {
        assert!(!self.binding.is_null());
        unsafe { &*self.binding }
    }

    pub fn is_auto_increment(&self) -> bool {
        self.is_auto_increment
    }

    pub fn get_binding_from_field(field: &Field<T>) -> &dyn TableBinding<T> {
        field.get_table_binding()
    }

    pub fn get_binding_from_fields<'a>(fields: &Vec<&'a Field<T>>) -> &'a dyn TableBinding<T> {
        assert!(!fields.is_empty());
        let field = fields[0];
        Self::get_binding_from_field(field)
    }
}
